
% Default to the notebook output style

    


% Inherit from the specified cell style.




    
\documentclass[11pt]{article}

    
    
    \usepackage[T1]{fontenc}
    % Nicer default font (+ math font) than Computer Modern for most use cases
    \usepackage{mathpazo}

    % Basic figure setup, for now with no caption control since it's done
    % automatically by Pandoc (which extracts ![](path) syntax from Markdown).
    \usepackage{graphicx}
    % We will generate all images so they have a width \maxwidth. This means
    % that they will get their normal width if they fit onto the page, but
    % are scaled down if they would overflow the margins.
    \makeatletter
    \def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth
    \else\Gin@nat@width\fi}
    \makeatother
    \let\Oldincludegraphics\includegraphics
    % Set max figure width to be 80% of text width, for now hardcoded.
    \renewcommand{\includegraphics}[1]{\Oldincludegraphics[width=.8\maxwidth]{#1}}
    % Ensure that by default, figures have no caption (until we provide a
    % proper Figure object with a Caption API and a way to capture that
    % in the conversion process - todo).
    \usepackage{caption}
    \DeclareCaptionLabelFormat{nolabel}{}
    \captionsetup{labelformat=nolabel}

    \usepackage{adjustbox} % Used to constrain images to a maximum size 
    \usepackage{xcolor} % Allow colors to be defined
    \usepackage{enumerate} % Needed for markdown enumerations to work
    \usepackage{geometry} % Used to adjust the document margins
    \usepackage{amsmath} % Equations
    \usepackage{amssymb} % Equations
    \usepackage{textcomp} % defines textquotesingle
    % Hack from http://tex.stackexchange.com/a/47451/13684:
    \AtBeginDocument{%
        \def\PYZsq{\textquotesingle}% Upright quotes in Pygmentized code
    }
    \usepackage{upquote} % Upright quotes for verbatim code
    \usepackage{eurosym} % defines \euro
    \usepackage[mathletters]{ucs} % Extended unicode (utf-8) support
    \usepackage[utf8x]{inputenc} % Allow utf-8 characters in the tex document
    \usepackage{fancyvrb} % verbatim replacement that allows latex
    \usepackage{grffile} % extends the file name processing of package graphics 
                         % to support a larger range 
    % The hyperref package gives us a pdf with properly built
    % internal navigation ('pdf bookmarks' for the table of contents,
    % internal cross-reference links, web links for URLs, etc.)
    \usepackage{hyperref}
    \usepackage{longtable} % longtable support required by pandoc >1.10
    \usepackage{booktabs}  % table support for pandoc > 1.12.2
    \usepackage[inline]{enumitem} % IRkernel/repr support (it uses the enumerate* environment)
    \usepackage[normalem]{ulem} % ulem is needed to support strikethroughs (\sout)
                                % normalem makes italics be italics, not underlines
    

    
    
    % Colors for the hyperref package
    \definecolor{urlcolor}{rgb}{0,.145,.698}
    \definecolor{linkcolor}{rgb}{.71,0.21,0.01}
    \definecolor{citecolor}{rgb}{.12,.54,.11}

    % ANSI colors
    \definecolor{ansi-black}{HTML}{3E424D}
    \definecolor{ansi-black-intense}{HTML}{282C36}
    \definecolor{ansi-red}{HTML}{E75C58}
    \definecolor{ansi-red-intense}{HTML}{B22B31}
    \definecolor{ansi-green}{HTML}{00A250}
    \definecolor{ansi-green-intense}{HTML}{007427}
    \definecolor{ansi-yellow}{HTML}{DDB62B}
    \definecolor{ansi-yellow-intense}{HTML}{B27D12}
    \definecolor{ansi-blue}{HTML}{208FFB}
    \definecolor{ansi-blue-intense}{HTML}{0065CA}
    \definecolor{ansi-magenta}{HTML}{D160C4}
    \definecolor{ansi-magenta-intense}{HTML}{A03196}
    \definecolor{ansi-cyan}{HTML}{60C6C8}
    \definecolor{ansi-cyan-intense}{HTML}{258F8F}
    \definecolor{ansi-white}{HTML}{C5C1B4}
    \definecolor{ansi-white-intense}{HTML}{A1A6B2}

    % commands and environments needed by pandoc snippets
    % extracted from the output of `pandoc -s`
    \providecommand{\tightlist}{%
      \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
    \DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
    % Add ',fontsize=\small' for more characters per line
    \newenvironment{Shaded}{}{}
    \newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.00,0.44,0.13}{\textbf{{#1}}}}
    \newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.56,0.13,0.00}{{#1}}}
    \newcommand{\DecValTok}[1]{\textcolor[rgb]{0.25,0.63,0.44}{{#1}}}
    \newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.25,0.63,0.44}{{#1}}}
    \newcommand{\FloatTok}[1]{\textcolor[rgb]{0.25,0.63,0.44}{{#1}}}
    \newcommand{\CharTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{{#1}}}
    \newcommand{\StringTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{{#1}}}
    \newcommand{\CommentTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textit{{#1}}}}
    \newcommand{\OtherTok}[1]{\textcolor[rgb]{0.00,0.44,0.13}{{#1}}}
    \newcommand{\AlertTok}[1]{\textcolor[rgb]{1.00,0.00,0.00}{\textbf{{#1}}}}
    \newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.02,0.16,0.49}{{#1}}}
    \newcommand{\RegionMarkerTok}[1]{{#1}}
    \newcommand{\ErrorTok}[1]{\textcolor[rgb]{1.00,0.00,0.00}{\textbf{{#1}}}}
    \newcommand{\NormalTok}[1]{{#1}}
    
    % Additional commands for more recent versions of Pandoc
    \newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.53,0.00,0.00}{{#1}}}
    \newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{{#1}}}
    \newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{{#1}}}
    \newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.73,0.40,0.53}{{#1}}}
    \newcommand{\ImportTok}[1]{{#1}}
    \newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.73,0.13,0.13}{\textit{{#1}}}}
    \newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{{#1}}}}}
    \newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{{#1}}}}}
    \newcommand{\VariableTok}[1]{\textcolor[rgb]{0.10,0.09,0.49}{{#1}}}
    \newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.00,0.44,0.13}{\textbf{{#1}}}}
    \newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.40,0.40,0.40}{{#1}}}
    \newcommand{\BuiltInTok}[1]{{#1}}
    \newcommand{\ExtensionTok}[1]{{#1}}
    \newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.74,0.48,0.00}{{#1}}}
    \newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.49,0.56,0.16}{{#1}}}
    \newcommand{\InformationTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{{#1}}}}}
    \newcommand{\WarningTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{{#1}}}}}
    
    
    % Define a nice break command that doesn't care if a line doesn't already
    % exist.
    \def\br{\hspace*{\fill} \\* }
    % Math Jax compatability definitions
    \def\gt{>}
    \def\lt{<}
    % Document parameters
    \title{pipeline\_exercise}
    
    
    

    % Pygments definitions
    
\makeatletter
\def\PY@reset{\let\PY@it=\relax \let\PY@bf=\relax%
    \let\PY@ul=\relax \let\PY@tc=\relax%
    \let\PY@bc=\relax \let\PY@ff=\relax}
\def\PY@tok#1{\csname PY@tok@#1\endcsname}
\def\PY@toks#1+{\ifx\relax#1\empty\else%
    \PY@tok{#1}\expandafter\PY@toks\fi}
\def\PY@do#1{\PY@bc{\PY@tc{\PY@ul{%
    \PY@it{\PY@bf{\PY@ff{#1}}}}}}}
\def\PY#1#2{\PY@reset\PY@toks#1+\relax+\PY@do{#2}}

\expandafter\def\csname PY@tok@w\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.73,0.73}{##1}}}
\expandafter\def\csname PY@tok@c\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@cp\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.74,0.48,0.00}{##1}}}
\expandafter\def\csname PY@tok@k\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@kp\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@kt\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.69,0.00,0.25}{##1}}}
\expandafter\def\csname PY@tok@o\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@ow\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.67,0.13,1.00}{##1}}}
\expandafter\def\csname PY@tok@nb\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@nf\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,1.00}{##1}}}
\expandafter\def\csname PY@tok@nc\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,1.00}{##1}}}
\expandafter\def\csname PY@tok@nn\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,1.00}{##1}}}
\expandafter\def\csname PY@tok@ne\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.82,0.25,0.23}{##1}}}
\expandafter\def\csname PY@tok@nv\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@no\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.53,0.00,0.00}{##1}}}
\expandafter\def\csname PY@tok@nl\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.63,0.63,0.00}{##1}}}
\expandafter\def\csname PY@tok@ni\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.60,0.60,0.60}{##1}}}
\expandafter\def\csname PY@tok@na\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.49,0.56,0.16}{##1}}}
\expandafter\def\csname PY@tok@nt\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@nd\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.67,0.13,1.00}{##1}}}
\expandafter\def\csname PY@tok@s\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@sd\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@si\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.73,0.40,0.53}{##1}}}
\expandafter\def\csname PY@tok@se\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.73,0.40,0.13}{##1}}}
\expandafter\def\csname PY@tok@sr\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.40,0.53}{##1}}}
\expandafter\def\csname PY@tok@ss\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@sx\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@m\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@gh\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,0.50}{##1}}}
\expandafter\def\csname PY@tok@gu\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.50,0.00,0.50}{##1}}}
\expandafter\def\csname PY@tok@gd\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.63,0.00,0.00}{##1}}}
\expandafter\def\csname PY@tok@gi\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.63,0.00}{##1}}}
\expandafter\def\csname PY@tok@gr\endcsname{\def\PY@tc##1{\textcolor[rgb]{1.00,0.00,0.00}{##1}}}
\expandafter\def\csname PY@tok@ge\endcsname{\let\PY@it=\textit}
\expandafter\def\csname PY@tok@gs\endcsname{\let\PY@bf=\textbf}
\expandafter\def\csname PY@tok@gp\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,0.50}{##1}}}
\expandafter\def\csname PY@tok@go\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.53,0.53,0.53}{##1}}}
\expandafter\def\csname PY@tok@gt\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.27,0.87}{##1}}}
\expandafter\def\csname PY@tok@err\endcsname{\def\PY@bc##1{\setlength{\fboxsep}{0pt}\fcolorbox[rgb]{1.00,0.00,0.00}{1,1,1}{\strut ##1}}}
\expandafter\def\csname PY@tok@kc\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@kd\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@kn\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@kr\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@bp\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@fm\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,1.00}{##1}}}
\expandafter\def\csname PY@tok@vc\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@vg\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@vi\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@vm\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@sa\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@sb\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@sc\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@dl\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@s2\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@sh\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@s1\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@mb\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@mf\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@mh\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@mi\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@il\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@mo\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@ch\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@cm\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@cpf\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@c1\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@cs\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}

\def\PYZbs{\char`\\}
\def\PYZus{\char`\_}
\def\PYZob{\char`\{}
\def\PYZcb{\char`\}}
\def\PYZca{\char`\^}
\def\PYZam{\char`\&}
\def\PYZlt{\char`\<}
\def\PYZgt{\char`\>}
\def\PYZsh{\char`\#}
\def\PYZpc{\char`\%}
\def\PYZdl{\char`\$}
\def\PYZhy{\char`\-}
\def\PYZsq{\char`\'}
\def\PYZdq{\char`\"}
\def\PYZti{\char`\~}
% for compatibility with earlier versions
\def\PYZat{@}
\def\PYZlb{[}
\def\PYZrb{]}
\makeatother


    % Exact colors from NB
    \definecolor{incolor}{rgb}{0.0, 0.0, 0.5}
    \definecolor{outcolor}{rgb}{0.545, 0.0, 0.0}



    
    % Prevent overflowing lines due to hard-to-break entities
    \sloppy 
    % Setup hyperref package
    \hypersetup{
      breaklinks=true,  % so long urls are correctly broken across lines
      colorlinks=true,
      urlcolor=urlcolor,
      linkcolor=linkcolor,
      citecolor=citecolor,
      }
    % Slightly bigger margins than the latex defaults
    
    \geometry{verbose,tmargin=1in,bmargin=1in,lmargin=1in,rmargin=1in}
    
    

    \begin{document}
    
    
    \maketitle
    
    

    
    \hypertarget{pipeline-exercise}{%
\section{Pipeline exercise}\label{pipeline-exercise}}

Fortunately there are computational pipelines which enable you to
process many samples jointly and which make the whole workflow more
user-friendly. These pipelines also help to produce a consistent,
documented and therefore reproducible workflow. Here we are going to use
the \href{https://github.com/AntonelliLab/seqcap_processor}{SECAPR
pipeline} on a dataset of \textbf{Ultraconserved Elements (UCEs)} that
were samples for the in South America.

    It's not clear if the existing morphological species assignments are
justified and if there might be cryptic species within these
morphospecies. We want to use this UCE dataset to generate a phylogeny
(species tree) of these samples to define coalescent species and see if
these assignments are in agreement with population genetics analyses
using SNP data extracted from the UCEs.

    \begin{center}\rule{0.5\linewidth}{\linethickness}\end{center}

\textbf{-1)} First you need to \textbf{update the SECAPR pipeline} by
pulling the newest version from GitHub and installing it in your cluster
environment. Simply execute the command below:

    \begin{Verbatim}[commandchars=\\\{\}]
{\color{incolor}In [{\color{incolor} }]:} \PYZpc{}\PYZpc{}bash
        sh /work/projects/forbio/bin/update\PYZus{}secapr.sh
\end{Verbatim}


    \begin{center}\rule{0.5\linewidth}{\linethickness}\end{center}

\textbf{0)} Let's first make sure that we are connected to the correct
software environment (\texttt{forbio\_env})

    \begin{Verbatim}[commandchars=\\\{\}]
{\color{incolor}In [{\color{incolor} }]:} \PYZpc{}\PYZpc{}bash
        module load Anaconda3/5.1.0
        \PY{n+nb}{source} activate forbio\PYZus{}env
\end{Verbatim}


    \begin{center}\rule{0.5\linewidth}{\linethickness}\end{center}

\textbf{1)} Then copy the pipeline tutorial folder into your directory
at \texttt{/work/users/USERNAME/}.

    \begin{Verbatim}[commandchars=\\\{\}]
{\color{incolor}In [{\color{incolor} }]:} \PYZpc{}\PYZpc{}bash
        cp \PYZhy{}r /work/projects/forbio/tutorials\PYZus{}tobi/pipeline/ /work/users/USERNAME/
\end{Verbatim}


    \begin{center}\rule{0.5\linewidth}{\linethickness}\end{center}

\textbf{2)} Now let's run the cleaning and trimming script for all of
your samples.

We are appending \texttt{2\textgreater{}\ warnings.txt} to all following
SECAPR commands because the cluster is printing a lot of annoying
warning messages when loading some of the SECAPR dependencies. This
command will silence those warnings and print them into the file
\texttt{warnings.txt}

Every command has a help function (\texttt{-h}) that shows you the
available options, which I always recommend to use before runnign a
function. Check out the help function of \texttt{secapr\ clean\_reads}:

    \begin{Verbatim}[commandchars=\\\{\}]
{\color{incolor}In [{\color{incolor} }]:} \PYZpc{}\PYZpc{}bash
        secapr clean\PYZus{}reads \PYZhy{}h \PY{l+m}{2}\PYZgt{} warnings.txt
\end{Verbatim}


    Now run the cleaning and trimming with this command. Feel free to add
any flags you feel are necessary.

    \begin{Verbatim}[commandchars=\\\{\}]
{\color{incolor}In [{\color{incolor} }]:} \PYZpc{}\PYZpc{}bash
        secapr clean\PYZus{}reads \PYZhy{}\PYZhy{}input raw\PYZus{}reads/ \PYZhy{}\PYZhy{}config helpfiles/adapters\PYZus{}info\PYZus{}topaza.txt \PYZhy{}\PYZhy{}output cleaned\PYZus{}reads \PYZhy{}\PYZhy{}index single \PYZhy{}\PYZhy{}headCrop \PY{l+m}{10} \PYZhy{}\PYZhy{}cropToLength \PY{l+m}{240} \PY{l+m}{2}\PYZgt{} warnings.txt
\end{Verbatim}


    Once it is running \textbf{INTERRUPT THIS COMMAND} using
\texttt{ctrl+c}. Since it will take around 30 minutes to clean all the
samples, we instead submit this as a job script. You find the job script
in the \texttt{scripts} folder. Fill in the correct paths and submit it
with:

    \begin{Verbatim}[commandchars=\\\{\}]
{\color{incolor}In [{\color{incolor} }]:} \PYZpc{}\PYZpc{}bash
        nano scripts/clean\PYZus{}trim\PYZus{}secapr.sh
        sbatch scripts/clean\PYZus{}trim\PYZus{}secapr.sh
\end{Verbatim}


    \textbf{While you are waiting} check out the
\href{https://phyluce.readthedocs.io/en/latest/}{PHYLUCE pipeline},
which is what you would normally use for UCE data. On the PHYLUCE page
you can also find
\href{https://phyluce.readthedocs.io/en/latest/tutorial-one.html}{cool
tutorials}, which you might want to play around with. Here we are using
the
\href{https://github.com/AntonelliLab/seqcap_processor/blob/master/documentation.ipynb}{SECAPR}
pipeline because it represents the workflow from yesterday's tutorial
and it gives a better handle for other, non-UCE target capture datasets.

    \begin{center}\rule{0.5\linewidth}{\linethickness}\end{center}

\textbf{3)} You can check the quality of the cleaned reads for all
samples using the \texttt{secapr\ quality\_check} command. This will
create a plot \texttt{QC\_plots.pdf} with an overview of the failed and
passed tests of all samples. It's okay for this exercise if the cleaned
read files are not perfect, since cleaning takes so long it's a bit of a
nuisance to get the settings right within the scope of this tutorial.

    \begin{Verbatim}[commandchars=\\\{\}]
{\color{incolor}In [{\color{incolor} }]:} \PYZpc{}\PYZpc{}bash
        secapr quality\PYZus{}check \PYZhy{}\PYZhy{}input cleaned\PYZus{}reads/ \PYZhy{}\PYZhy{}output quality\PYZus{}test \PY{l+m}{2}\PYZgt{} warnings.txt
\end{Verbatim}


    \begin{center}\rule{0.5\linewidth}{\linethickness}\end{center}

\textbf{4)} Now run a de novo assembly:

    \begin{Verbatim}[commandchars=\\\{\}]
{\color{incolor}In [{\color{incolor} }]:} \PYZpc{}\PYZpc{}bash
        secapr assemble\PYZus{}reads \PYZhy{}\PYZhy{}input ./cleaned\PYZus{}reads/ \PYZhy{}\PYZhy{}output ./contigs\PYZus{}abyss \PYZhy{}\PYZhy{}assembler abyss \PY{l+m}{2}\PYZgt{} warnings.txt
\end{Verbatim}


    This will take several hours. Luckily you can find already assembled
contigs for these samples on the cluster
(\texttt{/work/projects/forbio/data/contigs\_abyss/}), which you can
work with. Just copy the contigs to your working directory.

    \begin{Verbatim}[commandchars=\\\{\}]
{\color{incolor}In [{\color{incolor} }]:} \PYZpc{}\PYZpc{}bash
        cp \PYZhy{}r /work/projects/forbio/data/contigs\PYZus{}abyss/ .
\end{Verbatim}


    \begin{center}\rule{0.5\linewidth}{\linethickness}\end{center}

\textbf{5)} Extract the target regions:

    \begin{Verbatim}[commandchars=\\\{\}]
{\color{incolor}In [{\color{incolor} }]:} \PYZpc{}\PYZpc{}bash
        secapr find\PYZus{}target\PYZus{}contigs \PYZhy{}\PYZhy{}contigs contigs\PYZus{}abyss/ \PYZhy{}\PYZhy{}reference helpfiles/Tetrapods\PYZhy{}UCE\PYZhy{}2.5Kv1.fasta \PYZhy{}\PYZhy{}output target\PYZus{}contigs \PY{l+m}{2}\PYZgt{} warnings.txt
\end{Verbatim}


    Check the output folder and have a look at the \texttt{match\_table.txt}
file. Ask if you don't understand the information in the
\texttt{match\_table.txt} file.

Reminder: Always check the help function for every command before
running it by typing
\texttt{secapr\ NAME\_OF\_FUNCTION\ -h\ 2\textgreater{}\ warnings.txt}.

    \begin{center}\rule{0.5\linewidth}{\linethickness}\end{center}

\textbf{6)} Build multiple sequence alignments (MSAs) between all our
samples

    \begin{Verbatim}[commandchars=\\\{\}]
{\color{incolor}In [{\color{incolor} }]:} \PYZpc{}\PYZpc{}bash
        secapr align\PYZus{}sequences \PYZhy{}\PYZhy{}sequences target\PYZus{}contigs/extracted\PYZus{}target\PYZus{}contigs\PYZus{}all\PYZus{}samples.fasta \PYZhy{}\PYZhy{}output alignments/contig\PYZus{}alignments/ \PYZhy{}\PYZhy{}aligner mafft \PYZhy{}\PYZhy{}output\PYZhy{}format fasta \PYZhy{}\PYZhy{}no\PYZhy{}trim \PYZhy{}\PYZhy{}ambiguous \PY{l+m}{2}\PYZgt{} warnings.txt
\end{Verbatim}


    If you don't want to wait, \textbf{CANCEL} and run the script
\texttt{sbatch\ scripts/align\_contigs.sh} instead which is a lot faster
since it runs on parallel on 16 cores.

    Go to the output folder and check how many alignments were assembled.
You can use the command \texttt{ls\ \textbar{}\ wc\ -l} which gives you
the number of files in a folder. \_\_\_\_\_\_

\textbf{7)} Now we run the reference assembly, using our extracted
contigs as reference. We'll use the following command structure, but
will submit this as a job script
(\texttt{script/reference\_assembly.sh}), since it will take a little
while:

    \begin{Verbatim}[commandchars=\\\{\}]
{\color{incolor}In [{\color{incolor} }]:} \PYZpc{}\PYZpc{}bash
        secapr reference\PYZus{}assembly \PYZhy{}\PYZhy{}reads cleaned\PYZus{}reads \PYZhy{}\PYZhy{}reference\PYZus{}type alignment\PYZhy{}consensus \PYZhy{}\PYZhy{}reference alignments/contig\PYZus{}alignments \PYZhy{}\PYZhy{}output remapped\PYZus{}reads \PYZhy{}\PYZhy{}min\PYZus{}coverage \PY{l+m}{4} \PY{l+m}{2}\PYZgt{} warnings.txt
\end{Verbatim}


    If you are not entirely sure what this command does, check the
\href{https://github.com/AntonelliLab/seqcap_processor/blob/c4e9975629544f5b8b27ba02636effe1c4900d5d/docs/notebook/reference_assembly.ipynb}{SECAPR
manual} for an explanation of this command.

Submit the job script with:

    \begin{Verbatim}[commandchars=\\\{\}]
{\color{incolor}In [{\color{incolor} }]:} \PYZpc{}\PYZpc{}bash
        sbatch script/reference\PYZus{}assembly.sh
\end{Verbatim}


    What reference did we use for read mapping? Can you argue why our choice
is appropiate in this case (in case you agree)? Hint: All samples are
from the same genus (Topaza) and are thus relatively closely related.

Inspect some of the files using \texttt{samtools\ tview}.

\begin{center}\rule{0.5\linewidth}{\linethickness}\end{center}

\textbf{8)} You can use the \texttt{secapr\ locus\_selection} function
to find the loci that were assembled across all samples. This function
will find extract the \texttt{-n} loci with the best coverage across all
samples. It will only extract loci that are present in all samples. You
can set the number of loci to extract (\texttt{-n}) very high, to ensure
that all loci that are present in all samples will be extracted.

    \begin{Verbatim}[commandchars=\\\{\}]
{\color{incolor}In [{\color{incolor} }]:} \PYZpc{}\PYZpc{}bash
        secapr locus\PYZus{}selection \PYZhy{}\PYZhy{}input remapped\PYZus{}reads \PYZhy{}\PYZhy{}output locus\PYZus{}selection/ \PYZhy{}\PYZhy{}n \PY{l+m}{2500} \PYZhy{}\PYZhy{}read\PYZus{}cov \PY{l+m}{3} \PYZhy{}\PYZhy{}reference remapped\PYZus{}reads/reference\PYZus{}seqs/joined\PYZus{}fasta\PYZus{}library.fasta \PY{l+m}{2}\PYZgt{} warnings.txt
\end{Verbatim}


    \begin{center}\rule{0.5\linewidth}{\linethickness}\end{center}

\textbf{9)} Now we can build alignments from these loci:

    \begin{Verbatim}[commandchars=\\\{\}]
{\color{incolor}In [{\color{incolor} }]:} \PYZpc{}\PYZpc{}bash
        secapr align\PYZus{}sequences \PYZhy{}\PYZhy{}sequences locus\PYZus{}selection/joined\PYZus{}fastas\PYZus{}selected\PYZus{}loci.fasta \PYZhy{}\PYZhy{}output alignments/exon\PYZus{}intron\PYZus{}alignments/ \PYZhy{}\PYZhy{}aligner mafft \PYZhy{}\PYZhy{}output\PYZhy{}format fasta \PYZhy{}\PYZhy{}no\PYZhy{}trim \PYZhy{}\PYZhy{}ambiguous \PY{l+m}{2}\PYZgt{} warnings.txt
\end{Verbatim}


    What is the difference between these alignments and the ones we created
in step \textbf{6)}? Can you think of any reason why these alignments
might contain more loci for more samples?

\begin{center}\rule{0.5\linewidth}{\linethickness}\end{center}

\textbf{10)} SECAPR also has a function that enables allele phasing.
This will produce two separate BAM files per sample which in turn can be
summarized into two separate sequences (allele sequences) per sample and
locus:

    \begin{Verbatim}[commandchars=\\\{\}]
{\color{incolor}In [{\color{incolor} }]:} \PYZpc{}\PYZpc{}bash
        secapr phase\PYZus{}alleles \PYZhy{}\PYZhy{}input remapped\PYZus{}reads/ \PYZhy{}\PYZhy{}output allele\PYZus{}sequences \PYZhy{}\PYZhy{}min\PYZus{}coverage \PY{l+m}{3} \PYZhy{}\PYZhy{}reference remapped\PYZus{}reads/reference\PYZus{}seqs/joined\PYZus{}fasta\PYZus{}library.fasta
\end{Verbatim}


    \begin{center}\rule{0.5\linewidth}{\linethickness}\end{center}

\textbf{11)} Finally we can use the phased BAM files to generate allele
sequence alignments (MSAs) for all samples:

    \begin{Verbatim}[commandchars=\\\{\}]
{\color{incolor}In [{\color{incolor} }]:} \PYZpc{}\PYZpc{}bash
        secapr align\PYZus{}sequences \PYZhy{}\PYZhy{}sequences allele\PYZus{}sequences/joined\PYZus{}allele\PYZus{}fastas.fasta \PYZhy{}\PYZhy{}output alignments/allele\PYZus{}alignments/ \PYZhy{}\PYZhy{}aligner mafft \PYZhy{}\PYZhy{}output\PYZhy{}format fasta \PYZhy{}\PYZhy{}no\PYZhy{}trim \PYZhy{}\PYZhy{}ambiguous
\end{Verbatim}



    % Add a bibliography block to the postdoc
    
    
    
    \end{document}
